File: src/providers/zenvia/handlers.ts
import type { FastifyReply, FastifyRequest } from "fastify";
import { scheduleStatusAdvance } from "./scheduler.js";
import { createSmsMessageSchema, createSubscriptionSchema } from "./schemas.js";
import {
  createMessage,
  createSubscription,
  deleteSubscription,
  getSubscription,
  listSubscriptions,
} from "./state.js";

function toResponse(message: ReturnType<typeof createMessage>) {
  return {
    id: message.id,
    externalId: message.externalId ?? undefined,
    from: message.from,
    to: message.to,
    direction: message.direction,
    channel: message.channel,
    contents: message.contents,
    timestamp: message.createdAt,
  };
}

export async function createSmsMessageHandler(req: FastifyRequest, reply: FastifyReply) {
  const parsed = createSmsMessageSchema.safeParse(req.body);
  if (!parsed.success) {
    return reply.code(400).send({
      code: "VALIDATION_ERROR",
      message: "Validation error",
      details: parsed.error.issues.map((issue) => ({
        code: issue.code,
        path: issue.path.join("."),
        message: issue.message,
      })),
    });
  }

  const message = createMessage("sms", parsed.data);
  scheduleStatusAdvance(message.id);
  return reply.code(200).send(toResponse(message));
}

function toSubscriptionResponse(subscription: ReturnType<typeof createSubscription>) {
  return {
    id: subscription.id,
    eventType: subscription.eventType,
    webhook: {
      url: subscription.webhookUrl,
      headers: subscription.webhookHeaders ?? undefined,
    },
    criteria: {
      channel: subscription.criteriaChannel,
      direction: subscription.criteriaDirection ?? undefined,
    },
    status: subscription.status,
    version: "v2",
    createdAt: subscription.createdAt,
    updatedAt: subscription.updatedAt,
  };
}

export async function createSubscriptionHandler(req: FastifyRequest, reply: FastifyReply) {
  const parsed = createSubscriptionSchema.safeParse(req.body);
  if (!parsed.success) {
    return reply.code(400).send({
      code: "VALIDATION_ERROR",
      message: "Validation error",
      details: parsed.error.issues.map((issue) => ({
        code: issue.code,
        path: issue.path.join("."),
        message: issue.message,
      })),
    });
  }

  const subscription = createSubscription(parsed.data);
  return reply.code(200).send(toSubscriptionResponse(subscription));
}

export async function listSubscriptionsHandler(_req: FastifyRequest, reply: FastifyReply) {
  return reply.code(200).send({ content: listSubscriptions().map(toSubscriptionResponse) });
}

export async function getSubscriptionHandler(
  req: FastifyRequest<{ Params: { subscriptionId: string } }>,
  reply: FastifyReply,
) {
  const subscription = getSubscription(req.params.subscriptionId);
  if (!subscription) {
    return reply.code(404).send({ code: "NOT_FOUND", message: "Subscription not found" });
  }
  return reply.code(200).send(toSubscriptionResponse(subscription));
}

export async function deleteSubscriptionHandler(
  req: FastifyRequest<{ Params: { subscriptionId: string } }>,
  reply: FastifyReply,
) {
  const deleted = deleteSubscription(req.params.subscriptionId);
  if (!deleted) {
    return reply.code(404).send({ code: "NOT_FOUND", message: "Subscription not found" });
  }
  return reply.code(204).send();
}

File: internal/providers/zenvia/handlers.go
package zenvia

import (
	"encoding/json"
	"net/http"

	"github.com/golgimed/mimic/internal/shared/httpx"
)

func writeValidationError(w http.ResponseWriter) {
	httpx.WriteJSON(w, http.StatusBadRequest, map[string]any{
		"code":    "VALIDATION_ERROR",
		"message": "Validation error",
		"details": []map[string]string{{"path": "body", "message": "invalid or missing fields"}},
	})
}

type messageResponse struct {
	ID         string  `json:"id"`
	ExternalID *string `json:"externalId,omitempty"`
	From       string  `json:"from"`
	To         string  `json:"to"`
	Direction  string  `json:"direction"`
	Channel    string  `json:"channel"`
	Contents   []any   `json:"contents"`
	Timestamp  string  `json:"timestamp"`
}

func toMessageResponse(m *Message) messageResponse {
	return messageResponse{
		ID: m.ID, ExternalID: m.ExternalID, From: m.From, To: m.To,
		Direction: m.Direction, Channel: m.Channel, Contents: m.Contents, Timestamp: m.CreatedAt,
	}
}

func createMessageHandler(channel string, store *Store, scheduleAdvance func(string) error) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var wire createMessageWire
		if err := json.NewDecoder(r.Body).Decode(&wire); err != nil || !wire.valid(channel) {
			writeValidationError(w)
			return
		}

		message, err := store.CreateMessage(channel, wire.ExternalID, wire.From, wire.To, wire.toContents())
		if err != nil {
			httpx.WriteJSON(w, http.StatusInternalServerError, map[string]string{"code": "INTERNAL_ERROR", "message": err.Error()})
			return
		}

		if err := scheduleAdvance(message.ID); err != nil {
			httpx.WriteJSON(w, http.StatusInternalServerError, map[string]string{"code": "INTERNAL_ERROR", "message": err.Error()})
			return
		}

		httpx.WriteJSON(w, http.StatusOK, toMessageResponse(message))
	}
}

type subscriptionResponse struct {
	ID        string `json:"id"`
	EventType string `json:"eventType"`
	Webhook   struct {
		URL     string            `json:"url"`
		Headers map[string]string `json:"headers,omitempty"`
	} `json:"webhook"`
	Criteria struct {
		Channel   string  `json:"channel"`
		Direction *string `json:"direction,omitempty"`
	} `json:"criteria"`
	Status    string `json:"status"`
	Version   string `json:"version"`
	CreatedAt string `json:"createdAt"`
	UpdatedAt string `json:"updatedAt"`
}

func toSubscriptionResponse(s *Subscription) subscriptionResponse {
	var resp subscriptionResponse
	resp.ID = s.ID
	resp.EventType = s.EventType
	resp.Webhook.URL = s.WebhookURL
	resp.Webhook.Headers = s.WebhookHeaders
	resp.Criteria.Channel = s.CriteriaChannel
	resp.Criteria.Direction = s.CriteriaDirection
	resp.Status = s.Status
	resp.Version = "v2"
	resp.CreatedAt = s.CreatedAt
	resp.UpdatedAt = s.UpdatedAt
	return resp
}

func createSubscriptionHandler(store *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var wire createSubscriptionWire
		if err := json.NewDecoder(r.Body).Decode(&wire); err != nil || !wire.valid() {
			writeValidationError(w)
			return
		}

		sub, err := store.CreateSubscription(CreateSubscriptionInput{
			EventType:         wire.EventType,
			WebhookURL:        wire.Webhook.URL,
			WebhookHeaders:    wire.Webhook.Headers,
			CriteriaChannel:   wire.Criteria.Channel,
			CriteriaDirection: wire.Criteria.Direction,
		})
		if err != nil {
			httpx.WriteJSON(w, http.StatusInternalServerError, map[string]string{"code": "INTERNAL_ERROR", "message": err.Error()})
			return
		}

		httpx.WriteJSON(w, http.StatusOK, toSubscriptionResponse(sub))
	}
}

func listSubscriptionsHandler(store *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		subs, err := store.ListSubscriptions()
		if err != nil {
			httpx.WriteJSON(w, http.StatusInternalServerError, map[string]string{"code": "INTERNAL_ERROR", "message": err.Error()})
			return
		}
		content := make([]subscriptionResponse, len(subs))
		for i := range subs {
			content[i] = toSubscriptionResponse(&subs[i])
		}
		httpx.WriteJSON(w, http.StatusOK, map[string]any{"content": content})
	}
}

func getSubscriptionHandler(store *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		sub, err := store.GetSubscription(r.PathValue("subscriptionId"))
		if err != nil {
			httpx.WriteJSON(w, http.StatusInternalServerError, map[string]string{"code": "INTERNAL_ERROR", "message": err.Error()})
			return
		}
		if sub == nil {
			httpx.WriteJSON(w, http.StatusNotFound, map[string]string{"code": "NOT_FOUND", "message": "Subscription not found"})
			return
		}
		httpx.WriteJSON(w, http.StatusOK, toSubscriptionResponse(sub))
	}
}

func deleteSubscriptionHandler(store *Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		deleted, err := store.DeleteSubscription(r.PathValue("subscriptionId"))
		if err != nil {
			httpx.WriteJSON(w, http.StatusInternalServerError, map[string]string{"code": "INTERNAL_ERROR", "message": err.Error()})
			return
		}
		if !deleted {
			httpx.WriteJSON(w, http.StatusNotFound, map[string]string{"code": "NOT_FOUND", "message": "Subscription not found"})
			return
		}
		w.WriteHeader(http.StatusNoContent)
	}
}
